<!-- -->
<launch>
    <!-- frobit parameters -->
    <param name="diff_steer_wheel_distance" value="0.23"/>
    <param name="ticks_per_meter_left" value="562" />
    <param name="ticks_per_meter_right" value="562" />

    <!-- Notice: Use pkg="stage" if running ROS Groovy and pkg="stage_ros" if running ROS Hydro -->
    <node pkg="stage_ros" type="stageros" name="simulator" args="$(find frobit_demo)/stage/frobit_row_navigation.world"/>
    <remap from="/fmKnowledge/pose" to="/odom" />

    <group ns="fmSensors">
    </group>

    <group ns="fmProcessors">
		<node pkg="frobit_lego_transporter" name="wall_extractor" type="wall_extractor_node.py" output="screen">
		    <param name="laser_scan_sub" value="/base_scan"/>
		    <param name="ransac_debug_marker" value="/ransac_debug_marker" />
		    <param name="walls" value="/perception/walls" />
		</node>
    </group>

    <group ns="fmDecisionMakers">
        <node pkg="cross_bridge_mission" name="mission" type="cross_bridge_mission_node" output="screen">
            <param name="automode_sub" value="/fmPlan/automode"/>
            <param name="odom_sub" value="/odom" />
            <param name="straight_drive_pub" value="/straight_enb" />
            <param name="wall_follow_drive_pub" value="/wall_follow_enb" />
            <param name="dist_to_nonsafe_begin" value="0.8" />
            <param name="dist_to_nonsafe_end" value="1.1" />
            <param name="vel_cmd_pub" value="/fmCommand/vel_cmd" />
            <param name="disable_safety_pub" value="/disable_safety" />
        </node>
    </group>

    <group ns="fmExecutors">
		<node pkg="wall_follow_control" name="wall_follower" type="wall_follow_control_node" output="screen">
            <param name="drive_kp" value="2.0"/>
            <param name="drive_ki" value="0.1"/>
            <param name="drive_kd" value=".5"/>
            <param name="drive_feed_forward" value="0.00"/>
            <param name="drive_max_output" value="0.40"/>
			<param name="walls_sub" value="/perception/walls" />
			<param name="drive_max_i" value="1" />
			<param name="automode_sub" value="/wall_follow_enb" />
            <param name="command_pub" value="/fmCommand/cmd_vel" />
		</node>
    </group>

    <group ns="fmControllers">
    </group>

    <group ns="fmLib">
        <node pkg="twist_stamped_to_twist" name="converter" type="cmd_vel_converter.py">
            <param name="cmd_vel_sub" value="/safe_vel" />
            <param name="cmd_vel_pub" value="/cmd_vel" />
        </node>

        <node pkg="frobit_plot" name="frobit_plot" type="frobit_plot_node.py" output="screen" >
            <param name="pose_sub" value="/fmKnowledge/pose" />
            <param name="gnss_sub" value="/fmInformation/gpgga_tranmerc" />
            <param name="odom_sub" value="/fmKnowledge/odom" />
            <param name="wptnav_status_sub" value="/fmInformation/wptnav_status" />
            <param name="plot_pose_track" value="true" />
            <param name="plot_gnss_track" value="false" />
            <param name="plot_odometry_track" value="false" />
            <param name="plot_pose_yaw" value="false" />
            <param name="map_trackpoint_threshold" value="0.02" />
            <param name="map_max_trackpoints" value="5000" />
            <param name="map_minimum_size" value="2.0" />
            <param name="map_easting_offset" value="0.0" />
            <param name="map_northing_offset" value="0.0" />
            <param name="map_update_frequency" value="5.0" />
            <param name="map_title" value="Frobit track" />
            <param name="map_window_size" value="6.5" />
            <param name="avatar_extension_front" value="0.29" />
            <param name="avatar_extension_rear" value="0.06" />
            <param name="avatar_extension_lateral" value="0.13" />
            <param name="avatar_reverse" value="false" />
        </node>
    </group>

    <node pkg="drive_straight" name="drive_straight_node" type="drive_straight_node" output="screen">
        <param name="drive_speed" value="0.4" />
        <param name="straight_drive_enb_sub" value="/straight_enb" />
        <param name="straight_vel_pub" value="/fmCommand/cmd_vel" />        
    </node>

    <node pkg="safety_speed_node" name="obstacle_detector" type="lidar_obstacle_node.py" output="screen">
        <param name="reg_ratio_topic" value="/reg_ratio"/>
        <param name="scan_topic" value="/external/scan"/>
    </node>

    <node pkg="safety_speed_node" name="safety_speed" type="safety_speed_node.py" output="screen">
        <param name="vel_topic" value="/fmCommand/cmd_vel"/>
        <param name="safe_vel_topic" value="/safe_vel"/>
        <param name="reg_ratio_topic" value="/reg_ratio"/>
        <param name="disable_safety_topic" value="/disable_safety" />
    </node>

</launch>
